staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_serving_content.go

```diff
@@ -132,17 +134,54 @@ func (c *DynamicCertKeyPairContent) Run(workers int, stopCh <-chan struct{}) {
 	// doesn't matter what workers say, only start one.
 	go wait.Until(c.runWorker, time.Second, stopCh)
 
-	// start timer that rechecks every minute, just in case.  this also serves to prime the controller quickly.
-	go wait.PollImmediateUntil(FileRefreshDuration, func() (bool, error) {
-		c.queue.Add(workItemKey)
-		return false, nil
-	}, stopCh)
-
-	// TODO this can be wired to an fsnotifier as well.
+	// start the loop that watches the cert and key files until stopCh is closed.
+	go wait.Until(func() {
+		if err := c.watchCertKeyFile(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to watch cert and key file, will retry later")
+		}
+	}, time.Minute, stopCh)
 
 	<-stopCh
 }
 
+func (c *DynamicCertKeyPairContent) watchCertKeyFile(stopCh <-chan struct{}) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer w.Close()
+
+	if err = w.Add(c.certFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", c.certFile, err)
+	}
+	if err = w.Add(c.keyFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", c.keyFile, err)
+	}
+	// Reload the content when it starts a new watch as the file may have changed.
+	c.queue.Add(workItemKey)
+
+	for {
+		select {
+		case e := <-w.Events:
+			c.queue.Add(workItemKey)
+			// The original watched file is removed or renamed, restarting a new watch.
+			if e.Op&(fsnotify.Remove|e.Op&fsnotify.Rename) > 0 {
+				if err := w.Remove(e.Name); err != nil {
+					klog.InfoS("Failed to remove file watch, it may have been deleted", "file", e.Name, "err", err)
+				}
+				if err := w.Add(e.Name); err != nil {
+					return fmt.Errorf("error adding watch for file %s: %v", e.Name, err)
+				}
+				c.queue.Add(workItemKey)
```

Comment:
remind me how cert/key mismatches are handled.  A code reference comment right here about how it will handled when one is updated before the other would be helpful.  Previously, at a one minute polling cycle it was pretty unlikely, but I'm hoping past-me did something clever.

Reply:
You definitely handled this well: the certificate and key are validated before they are used. If either is missing or they don't match, an error will be returned and workqueue will reschedule it later with rate limiter. I will add comment to explain it.
https://github.com/kubernetes/kubernetes/blob/f1c817614d28ca1127a6217fce9032f37a4eec72/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_serving_content.go#L79-L96